src/text/raw_text_reader.rs

```diff
@@ -99,25 +70,30 @@ impl<T: ToIonDataSource> IonReader for RawTextReader<T> {
     }
 
     fn next(&mut self) -> IonResult<RawStreamItem> {
-        self.feed_if_needed()?;
-
-        match self.reader.next() {
-            Err(err @ IonError::IncompleteText { .. }) => {
-                // We received an incomplete.. and need more data..
-                loop {
-                    // If there is no more data to read, we need to bubble the incomplete up to the
-                    // caller.
-                    if 0 == self.read_source(1024)? {
-                        return Err(err);
-                    }
+        let mut read_size = self.max_read_size;
 
-                    match self.reader.next() {
-                        Err(IonError::IncompleteText { .. }) => (),
-                        other => return other,
+        loop {
+            let result = self.reader.next();
+            if let Err(IonError::IncompleteText { .. }) = result {
+                let bytes_read = self.read_source(read_size)?;
+                // if we have no bytes, and our stream has been marked as fully loaded, then we
+                // need to bubble up the error. Otherwise, if our stream has not been marked as
+                // loaded, then we need to mark it as loaded and retry.
+                if 0 == bytes_read {
+                    if self.reader.is_stream_done() {
+                        return result;
+                    } else {
+                        self.reader.stream_done();
                     }
                 }
+                // The assumption here is that most buffer sizes will start at a magnitude the user
+                // is comfortable with in terms of memory usage. So if we're reading more in order
+                // to reach a parsable point we do not want to start consuming more than an order of
+                // magnitude more memory just to get there.
+                read_size = std::cmp::min(read_size * 2, self.max_read_size * 10);
```

Comment:
We're willing to read by 10x the `max_read_size`, so I think we should reconsider that variable name. `original_buffer_size`, maybe?